processing/src/main/java/io/druid/segment/data/BlockLayoutSerde.java

```diff
@@ -0,0 +1,309 @@
+      {
+        return "BlockCompressedIndexedLongs_Anonymous{" +
+               "currIndex=" + currIndex +
+               ", sizePer=" + sizePer +
+               ", numChunks=" + singleThreadedLongBuffers.size() +
+               ", totalSize=" + totalSize +
+               '}';
+      }
+
+      @Override
+      public void close() throws IOException
+      {
+        Closeables.close(holder, false);
+      }
+    }
+  }
+
+
+  public static class BlockLayoutLongSupplierSerializer implements LongSupplierSerializer {
+
+    private final IOPeon ioPeon;
+    private final int sizePer;
+    private final CompressionFactory.LongEncodingFormatWriter writer;
+    private final GenericIndexedWriter<ResourceHolder<ByteBuffer>> flattener;
+    private final CompressedObjectStrategy.CompressionStrategy compression;
+    private final String metaFile;
+    private CountingOutputStream metaOut;
+
+    private int numInserted = 0;
+
+    private ByteBuffer endBuffer = null;
+
+    public BlockLayoutLongSupplierSerializer(
+        IOPeon ioPeon,
+        String filenameBase,
+        ByteOrder order,
+        CompressionFactory.LongEncodingFormatWriter writer,
+        CompressedObjectStrategy.CompressionStrategy compression
+    )
+    {
+      this.ioPeon = ioPeon;
+      this.sizePer = writer.getBlockSize(CompressedPools.BUFFER_SIZE);
+      this.flattener = new GenericIndexedWriter<ResourceHolder<ByteBuffer>>(
+                            ioPeon,
+                            filenameBase,
+                            VSizeCompressedObjectStrategy.getBufferForOrder(
+                                order,
+                                compression,
+                                writer.getNumBytes(sizePer)
+                            )
+                        );
+      this.metaFile = filenameBase + ".format";
+      this.writer = writer;
+      this.compression = compression;
+    }
+
+    public void open() throws IOException
+    {
+      flattener.open();
+    }
+
+    public int size()
+    {
+      return numInserted;
+    }
+
+    public void add(long value) throws IOException
+    {
+      if (numInserted % sizePer == 0) {
+        if (endBuffer != null) {
+          writer.close();
+          endBuffer.limit(endBuffer.position());
+          endBuffer.rewind();
+          flattener.write(StupidResourceHolder.create(endBuffer));
+        }
+        endBuffer = ByteBuffer.allocate(writer.getNumBytes(sizePer));
+        writer.setBuffer(endBuffer);
+      }
+
+      writer.write(value);
+      ++numInserted;
+    }
+
+    public void closeAndConsolidate(ByteSink consolidatedOut) throws IOException
+    {
+      close();
+      try (OutputStream out = consolidatedOut.openStream();
+           InputStream meta = ioPeon.makeInputStream(metaFile)) {
+        ByteStreams.copy(meta, out);
+        flattener.combineStreams().copyTo(out);
+      }
+    }
+
+    public void close() throws IOException {
+      if (endBuffer != null) {
```

Comment:
The previous CompressedLongSupplierSerializer had these methods for close. It seems that IndexMerger use closeAndConsolidate() to combine the data, while IndexMergerV9 use close() followed by writeToChannel().